### PR TITLE
[#144] Handle adapter errors 

### PR DESCRIPTION
**Pull request by pmeinhardt**
Okay. So after a long wait, here's a first take on notifying users when something goes wrong while scanning for tickets. I have an idea for even better error reporting which I'll try out – but that'll be a separate PR.

preview

### CODE REVIEW

**Comment by pmeinhardt** [comment]
Once this is in, I think we're good to go for a new release 🎉 
`3.0` since we now have API-based adapters?

File: src/common/popup/components/no-tickets.jsx [hunk 1]
```diff
@@ -20,12 +29,47 @@ function NoTickets() {
...
+
+function Report({ errors }) {
+  return (
+    <>
+      <h5 className="mt-3 mb-3">Houston, we have a problem.</h5>
+      <p>
+        Tickety-Tick encountered
+        {' '}
+        <strong>
+          {errors.length === 1 ? 'an error' : `${errors.length} errors`}
```
Comment (klappradla) [inline src/common/popup/components/no-tickets.jsx]: 👍 makes sense!

File: src/common/popup/components/tool.jsx [hunk 2]
```diff
@@ -4,25 +4,26 @@ import PropTypes from 'prop-types';
...
 import NoTickets from './no-tickets';
 import Header from './header';
+
 import TicketShape from '../utils/ticket-shape';
+import ErrorShape from '../utils/error-shape';
 
-function Tool({ tickets }) {
-  const content = (tickets && tickets.length > 0)
+function Tool({ tickets, errors }) {
+  const content = tickets.length > 0
```
Comment (klappradla) [inline src/common/popup/components/tool.jsx]: Perfect! So we're only ever considering the errors if there are not tickets? 👍

File: src/common/search.js [hunk 3]
```diff
@@ -5,11 +5,30 @@ import Ora from './adapters/ora';
...
+async function attempt(scan, loc, doc) {
+  try {
+    const tickets = await scan(loc, doc);
+    return { tickets };
+  } catch (error) {
+    return { error: serializable(error) };
+  }
+}
+
+function aggregate(results) {
```
Comment (klappradla) [inline src/common/search.js]: 💚 for the naming

File: src/common/utils/serializable-errors.test.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import serializable from './serializable-errors';
+
+describe('serializable-errors', () => {
+  const error = new Error('Serialized Test Error');
+
+  let representation;
+
+  function roundtrip(value) {
```
Comment (klappradla) [inline src/common/utils/serializable-errors.test.js]: Interesting approach 🤓

File: src/common/popup/components/tool.jsx (hunk 2 above)
Comment (pmeinhardt) [reply src/common/popup/components/tool.jsx]: Exactly. Our idea from the last hacknight was that as long as a matching adapter finds tickets, it's okay for other adapters to reject with errors. So only if we don't receive any tickets, but we see errors, we report them.

File: src/common/utils/serializable-errors.test.js (hunk 4 above)
Comment (pmeinhardt) [reply src/common/utils/serializable-errors.test.js]: Haha 😁 Yeah, I was wondering why I couldn't retrieve the error messages and stack traces in the pop-up window for a bit before it hit me… when they're passed from the content script to the pop-up window, they need to be serializable.

```javascript
JSON.stringify(new Error('Boom!')) === '{}'
```

**Pull request merged**